File: packages/research-backend/src/research_backend/chats/providers/claude.py
"""Claude export normalization."""

from __future__ import annotations

import hashlib
from typing import Any

from ..models import ChatAttachment, ChatConversation, ChatMessage


def parse_claude_export(data: Any) -> list[ChatConversation]:
    if isinstance(data, list):
        conversations = data
    elif isinstance(data, dict):
        conversations = data.get("chat_messages") or []
    else:
        conversations = []
    normalized: list[ChatConversation] = []

    for conv in conversations:
        raw_messages = conv.get("messages") or conv.get("chat_messages") or []
        messages = [
            ChatMessage(
                role=_normalize_role(message.get("role") or message.get("sender", "")),
                content=_extract_message_text(message),
                timestamp=str(message.get("timestamp") or message.get("created_at") or ""),
                model=str(message.get("model", "")),
                attachments=_parse_attachments(message.get("attachments")),
            )
            for message in raw_messages
        ]
        conversation_id = str(conv.get("id") or conv.get("uuid") or "").strip() or _fallback_conversation_id(conv, messages)
        normalized.append(
            ChatConversation(
                provider="claude",
                conversation_id=conversation_id,
                canonical_id=f"claude-{conversation_id}",
                title=str(conv.get("title") or conv.get("name") or "(untitled)"),
                created_at=str(conv.get("created_at", "")),
                updated_at=str(conv.get("updated_at", "")),
                assistant_models=_collect_assistant_models(messages),
                messages=messages,
            )
        )

    return normalized


def _normalize_role(role: Any) -> str:
    value = str(role or "").strip().lower()
    if value == "human":
        return "user"
    if value in {"assistant", "user", "system", "tool"}:
        return value
    return value or "user"


def _collect_assistant_models(messages: list[ChatMessage]) -> list[str]:
    models: list[str] = []
    for message in messages:
        if message.role != "assistant":
            continue
        if message.model and message.model not in models:
            models.append(message.model)
    return models


def _fallback_conversation_id(conv: dict[str, Any], messages: list[ChatMessage]) -> str:
    digest = hashlib.sha256()
    digest.update(str(conv.get("title", "")).encode("utf-8"))
    for message in messages[:3]:
        digest.update(message.role.encode("utf-8"))
        digest.update(message.timestamp.encode("utf-8"))
        digest.update(message.content.encode("utf-8"))
    return digest.hexdigest()[:12]


def _parse_attachments(raw_attachments: Any) -> list[ChatAttachment]:
    if not isinstance(raw_attachments, list):
        return []

    attachments: list[ChatAttachment] = []
    for attachment in raw_attachments:
        if not isinstance(attachment, dict):
            continue
        attachments.append(
            ChatAttachment(
                kind=str(
                    attachment.get("kind")
                    or attachment.get("type")
                    or "file"
                ),
                name=str(
                    attachment.get("name")
                    or attachment.get("file_name")
                    or ""
                ),
                status=str(attachment.get("status", "referenced_only") or "referenced_only"),
                mime_type=str(
                    attachment.get("mime_type")
                    or attachment.get("file_type")
                    or ""
                ),
                url=str(
                    attachment.get("url")
                    or attachment.get("preview_url")
                    or ""
                ),
            )
        )
    return attachments


def _extract_message_text(message: dict[str, Any]) -> str:
    text = str(message.get("text") or "").strip()
    if text:
        return text
    content = message.get("content")
    if not isinstance(content, list):
        return ""
    parts: list[str] = []
    for item in content:
        if not isinstance(item, dict):
            continue
        if item.get("type") == "text" and item.get("text"):
            parts.append(str(item["text"]))
        elif item.get("type") == "thinking" and item.get("thinking"):
            parts.append(str(item["thinking"]))
    return "\n\n".join(parts)

File: packages/ooda/src/imports/parsers/claude.ts
import { z } from "zod";

import type { ImportedConversation, ImportedMessage } from "../types.js";

// --- Zod schemas matching Claude Projects JSON export ---

const ClaudeContentPartSchema = z.object({
  type: z.string(),
  text: z.string().optional(),
  thinking: z.string().optional(),
});

const ClaudeMessageSchema = z.object({
  role: z.string().optional(),
  sender: z.string().optional(),
  text: z.string().optional(),
  content: z.array(ClaudeContentPartSchema).optional(),
  timestamp: z.union([z.string(), z.number()]).optional(),
  created_at: z.string().optional(),
  model: z.string().optional(),
});

const ClaudeConversationSchema = z.object({
  id: z.string().optional(),
  uuid: z.string().optional(),
  title: z.string().optional(),
  name: z.string().optional(),
  created_at: z.string().optional(),
  updated_at: z.string().optional(),
  messages: z.array(ClaudeMessageSchema).optional(),
  chat_messages: z.array(ClaudeMessageSchema).optional(),
});

const ClaudeExportSchema = z.union([
  z.array(ClaudeConversationSchema),
  z.object({ chat_messages: z.array(ClaudeConversationSchema).optional() }),
]);

// --- Helpers ---

function normalizeRole(role: string): ImportedMessage["role"] {
  const value = (role || "").trim().toLowerCase();
  if (value === "human") return "user";
  if (value === "assistant" || value === "system") return value;
  return "user";
}

function extractMessageText(msg: z.infer<typeof ClaudeMessageSchema>): string {
  const text = (msg.text ?? "").trim();
  if (text) return text;

  if (!Array.isArray(msg.content)) return "";
  const parts: string[] = [];
  for (const item of msg.content) {
    if (item.type === "text" && item.text) parts.push(item.text);
    else if (item.type === "thinking" && item.thinking) parts.push(item.thinking);
  }
  return parts.join("\n\n");
}

// --- Main parser ---

export function parseClaude(data: unknown): ImportedConversation[] {
  const parsed = ClaudeExportSchema.safeParse(data);
  if (!parsed.success) return [];

  const conversations = Array.isArray(parsed.data)
    ? parsed.data
    : parsed.data.chat_messages ?? [];

  const result: ImportedConversation[] = [];

  for (const conv of conversations) {
    const rawMessages = conv.messages ?? conv.chat_messages ?? [];
    const messages: ImportedMessage[] = [];

    for (const msg of rawMessages) {
      const content = extractMessageText(msg);
      if (!content) continue;
      messages.push({
        role: normalizeRole(msg.role ?? msg.sender ?? ""),
        content,
        timestamp: String(msg.timestamp ?? msg.created_at ?? ""),
        model: msg.model ?? undefined,
      });
    }

    const conversationId =
      (conv.id ?? conv.uuid ?? "").trim() || `claude-${result.length}`;

    result.push({
      provider: "claude",
      conversationId,
      title: conv.title ?? conv.name ?? "(untitled)",
      messages,
      createdAt: conv.created_at ?? undefined,
      updatedAt: conv.updated_at ?? undefined,
    });
  }

  return result;
}
